### PR TITLE
Add validation at the beginning of func:RandomSamplingWithPriority

### PR DESCRIPTION
**Pull request by Mdaiki0730**
## Description

I think we need adding independent input validation of the parameters at the beginning of the function RandomSamplingWithPriority() instead of relying on the selection algorithm to validate them.

## PR Check List
- [ ] Is this enough for validate type?
- [ ] Won't it cause a predictable critical panic?

### CODE REVIEW

File: libs/rand/sampling.go [hunk 1]
```diff
@@ -24,9 +24,16 @@ type Candidate interface {
 // Returns:
 // samples - A randomly selected candidate from a set of candidates. NOTE that the same candidate may have been
 // selected in duplicate.
+// TODO: Change to a function that returns an error, as it is not desirable for the function to end with a panic.
```
Comment (tnasu) [inline libs/rand/sampling.go:27]: Why do you leave TODO?

File: libs/rand/sampling.go [hunk 2]
```diff
@@ -50,15 +57,10 @@ func RandomSamplingWithPriority(
...
-	actualTotalPriority := uint64(0)
-	for i := 0; i < len(candidates); i++ {
-		actualTotalPriority += candidates[i].Priority()
-	}
-	panic(fmt.Sprintf("Either the given candidate is an empty set, the actual cumulative priority is zero,"+
-		" or the total priority is less than the actual one; totalPriority=%d, actualTotalPriority=%d,"+
-		" seed=%d, sampleSize=%d, undrawn=%d, threshold[%d]=%d, len(candidates)=%d",
-		totalPriority, actualTotalPriority, seed, sampleSize, undrawn, undrawn, thresholds[undrawn], len(candidates)))
+	// We're assuming you never get to this code
+	panic(fmt.Sprintf("An unexpected error occurred while selecting samples; "+
```
Comment (tnasu) [inline libs/rand/sampling.go:61]: Would you modify the error message to like `cannot select samples`?

**Review by tnasu — CHANGES_REQUESTED** [review]

File: libs/rand/sampling.go (hunk 1 above)
Comment (Mdaiki0730) [reply libs/rand/sampling.go:27]: Currently, panic is returned, and I thought that returning error as a libs function would be more versatile.
I'll fix it if it is wrong as how to leave TODO

File: libs/rand/sampling.go [hunk 3]
```diff
@@ -93,3 +95,28 @@ func nextRandom(rand *uint64) uint64 {
...
 }
+
+func checkInvalidPriority(candidates []Candidate, totalPriority uint64) error {
+	actualTotalPriority := uint64(0)
+	for i := 0; i < len(candidates); i++ {
+		actualTotalPriority += candidates[i].Priority()
+	}
+
+	if len(candidates) == 0 {
+		return fmt.Errorf("the specified candidate is an empty set; "+
```
Comment (ulbqb) [inline libs/rand/sampling.go:106]: candidates set is empty

File: libs/rand/sampling.go [hunk 4]
```diff
@@ -93,3 +95,28 @@ func nextRandom(rand *uint64) uint64 {
...
+			"totalPriority=%d, actualTotalPriority=%d, len(candidates)=%d",
+			totalPriority, actualTotalPriority, len(candidates))
+
+	} else if totalPriority == 0 || actualTotalPriority == 0 {
+		return fmt.Errorf("actual cumulative priority is zero; "+
```
Comment (ulbqb) [inline libs/rand/sampling.go:111]: total priority is zero

File: libs/rand/sampling.go (hunk 1 above)
Comment (tnasu) [reply libs/rand/sampling.go:27]: Then, it would be better to fix it instead of leaving `TODO`. I agree with your thoughts. How about @torao?

Comment (tnasu) [reply libs/rand/sampling.go:27]: IMO, `TODO` can stay forever.

**Review by ulbqb — COMMENTED** [review]
What about simple error message?

File: libs/rand/sampling.go (hunk 1 above)
Comment (Mdaiki0730) [reply libs/rand/sampling.go]: It's certainly not something that should be left in the code.
I'll remove it and leave it in the other document.

Comment (tnasu) [reply libs/rand/sampling.go]: Please make an issue. Thanks

**Review by tnasu — APPROVED** [review]

**Comment by Mdaiki0730** [comment]
@ulbqb 
Thank you for your suggestion. I just translated it directly into English, so I'll reflect your simpler comment.

**Review by tnasu — APPROVED** [review]

**Review by ulbqb — APPROVED** [review]

**Pull request merged**